File: _includes/sort-benchmark/algorithms/flash.rs
fn flash_class_index(x: usize, min: usize, max: usize, m: usize) -> usize {
    if max == min {
        0
    } else {
        ((x - min) as f64 / (max - min) as f64 * (m - 1) as f64) as usize
    }
}

fn flash_sort(a: &mut [usize]) {
    let n = a.len();
    if n <= 1 {
        return;
    }

    let min = *a.iter().min().unwrap();
    let max = *a.iter().max().unwrap();
    if min == max {
        return;
    }

    let m = ((n as f64 * (2.0 * n as f64).log2()).sqrt().ceil() as usize).clamp(2, n);

    let mut count = vec![0usize; m];
    for &x in a.iter() {
        count[flash_class_index(x, min, max, m)] += 1;
    }

    let mut boundary = vec![0usize; m + 1];
    for i in 0..m {
        boundary[i + 1] = boundary[i] + count[i];
    }

    let mut temp = vec![0usize; n];
    let mut cursor = boundary.clone();
    for &x in a.iter() {
        let k = flash_class_index(x, min, max, m);
        temp[cursor[k]] = x;
        cursor[k] += 1;
    }
    a.copy_from_slice(&temp);

    for i in 0..m {
        let start = boundary[i];
        let end = boundary[i + 1];
        if end - start > 1 {
            insertion_sort(&mut a[start..end]);
        }
    }
}

File: _includes/sort-benchmark/algorithms/flash.swift
private func flash_class_index(_ x: Int, _ minVal: Int, _ maxVal: Int, _ m: Int) -> Int {
    if maxVal == minVal {
        return 0
    } else {
        return Int(Double(x - minVal) / Double(maxVal - minVal) * Double(m - 1))
    }
}

func flash_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { flash_sort($0) }
}

func flash_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    let n = a.count
    if n <= 1 {
        return
    }

    var lo = a[0]
    var hi = a[0]
    for i in 1..<n {
        if a[i] < lo { lo = a[i] }
        if a[i] > hi { hi = a[i] }
    }
    if lo == hi {
        return
    }

    let mRaw = Int((Double(n) * log2(2.0 * Double(n))).squareRoot().rounded(.up))
    let m = Swift.min(Swift.max(mRaw, 2), n)

    var count = [Int](repeating: 0, count: m)
    for i in 0..<n {
        count[flash_class_index(a[i], lo, hi, m)] += 1
    }

    var boundary = [Int](repeating: 0, count: m + 1)
    for i in 0..<m {
        boundary[i + 1] = boundary[i] + count[i]
    }

    var temp = [Int](repeating: 0, count: n)
    var cursor = boundary
    for i in 0..<n {
        let k = flash_class_index(a[i], lo, hi, m)
        temp[cursor[k]] = a[i]
        cursor[k] += 1
    }
    for i in 0..<n {
        a[i] = temp[i]
    }

    for i in 0..<m {
        let start = boundary[i]
        let end = boundary[i + 1]
        if end - start > 1 {
            insertion_sort(UnsafeMutableBufferPointer(rebasing: a[start..<end]))
        }
    }
}
